include/crow/http_connection.h

```diff
@@ -447,6 +461,26 @@ namespace crow
             }
 
             buffers_.emplace_back(crlf.data(), crlf.size());
+            
+        }
+#if !defined(_WIN32)
+        void do_write_static(){
+            res.adaptor = &adaptor_;
```

Comment:
Type mismatch when assigning (`adaptor_` is template type). Maybe pass this variable as a parameter of `res.do_write_sendfile`?

Reply:
It looks like this solution doesn't work very well with SSL socket.

Reply:
this is gonna sound stupid probably, but I'm trying to use the `Adaptor` typename to solve the mismatch (i know i need to figure out the socket calls as well), but i keep getting Adaptor does not name a type

Reply:
You could pass adaptor as parameter to function and make it template. But it doesn't solve the problem - `adaptor->socket().native_handle()` for SSL socket doesn't return int.

Reply:
Yeah for SSL `adaptor->socket().lowest_layer().native_handle()` should do the trick.

Reply:
I found a method (`raw_socket()`) that returns the proper socket needed for both, and I'm passing the adapter as an argument, compiler isn't complaining about it anymore, only problem now is I can't compile with SSL enabled even without this PR, a bunch of undefined references to asio ssl methods, `SSL_connect`,`SSL_accept`,`SSL_read`,`SSL_write` + a bunch others.


```
g++ main.cpp -ggdb -o testexec -lpthread -lboost_system -D CROW_ENABLE_SSL
/usr/bin/ld: /usr/bin/ld: DWARF error: could not find variable specification at offset 6b32e
/usr/bin/ld: DWARF error: could not find variable specification at offset 70a40
/usr/bin/ld: DWARF error: could not find variable specification at offset 7e251
/usr/bin/ld: DWARF error: could not find variable specification at offset 71587
/usr/bin/ld: DWARF error: could not find variable specification at offset 846d9
/usr/bin/ld: DWARF error: could not find variable specification at offset 7296d
/usr/bin/ld: DWARF error: could not find variable specification at offset 7e47d
/usr/bin/ld: DWARF error: could not find variable specification at offset 7e3bb
/usr/bin/ld: DWARF error: could not find variable specification at offset 7e025
/usr/bin/ld: DWARF error: could not find variable specification at offset 7e306
/usr/bin/ld: DWARF error: could not find variable specification at offset 7e0da
/usr/bin/ld: DWARF error: could not find variable specification at offset 6b59d
/usr/bin/ld: DWARF error: could not find variable specification at offset 8c33e
/usr/bin/ld: DWARF error: could not find variable specification at offset 8c34c
/usr/bin/ld: DWARF error: could not find variable specification at offset 72960
/tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::openssl_init_base::do_init::~do_init()':
/usr/include/boost/asio/ssl/detail/impl/openssl_init.ipp:90: undefined reference to `CONF_modules_unload'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::error::detail::ssl_category::message[abi:cxx11](int) const':
/usr/include/boost/asio/ssl/impl/error.ipp:39: undefined reference to `ERR_reason_error_string'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::context::context(boost::asio::ssl::context_base::method)':
/usr/include/boost/asio/ssl/impl/context.ipp:66: undefined reference to `ERR_clear_error'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:93: undefined reference to `TLS_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:93: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:96: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:97: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:101: undefined reference to `TLS_client_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:101: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:104: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:105: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:109: undefined reference to `TLS_server_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:109: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:112: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:113: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:138: undefined reference to `TLS_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:138: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:141: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:142: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:146: undefined reference to `TLS_client_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:146: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:149: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:150: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:154: undefined reference to `TLS_server_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:154: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:157: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:158: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:183: undefined reference to `TLS_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:183: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:186: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:187: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:191: undefined reference to `TLS_client_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:191: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:194: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:195: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:199: undefined reference to `TLS_server_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:199: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:202: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:203: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:228: undefined reference to `TLS_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:228: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:231: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:232: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:236: undefined reference to `TLS_client_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:236: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:239: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:240: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:244: undefined reference to `TLS_server_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:244: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:247: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:248: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:274: undefined reference to `TLS_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:274: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:277: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:278: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:282: undefined reference to `TLS_client_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:282: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:285: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:286: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:290: undefined reference to `TLS_server_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:290: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:293: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:294: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:310: undefined reference to `TLS_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:310: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:313: undefined reference to `TLS_client_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:313: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:316: undefined reference to `TLS_server_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:316: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:322: undefined reference to `TLS_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:322: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:324: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:327: undefined reference to `TLS_client_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:327: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:329: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:332: undefined reference to `TLS_server_method'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:332: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:334: undefined reference to `SSL_CTX_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:355: undefined reference to `SSL_CTX_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:362: undefined reference to `ERR_get_error'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::context::~context()':
/usr/include/boost/asio/ssl/impl/context.ipp:393: undefined reference to `SSL_CTX_get_default_passwd_cb_userdata'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:406: undefined reference to `SSL_CTX_set_default_passwd_cb_userdata'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:412: undefined reference to `SSL_CTX_get_ex_data'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:416: undefined reference to `SSL_CTX_get_ex_data'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:418: undefined reference to `SSL_CTX_set_ex_data'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:421: undefined reference to `SSL_CTX_free'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::context::set_options(long, boost::system::error_code&)':
/usr/include/boost/asio/ssl/impl/context.ipp:485: undefined reference to `SSL_CTX_set_options'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::context::set_verify_mode(int, boost::system::error_code&)':
/usr/include/boost/asio/ssl/impl/context.ipp:501: undefined reference to `SSL_CTX_get_verify_callback'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:501: undefined reference to `SSL_CTX_set_verify'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::context::use_certificate_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, boost::asio::ssl::context_base::file_format, boost::system::error_code&)':
/usr/include/boost/asio/ssl/impl/context.ipp:713: undefined reference to `ERR_clear_error'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:715: undefined reference to `SSL_CTX_use_certificate_file'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:718: undefined reference to `ERR_get_error'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::context::use_private_key_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, boost::asio::ssl::context_base::file_format, boost::system::error_code&)':
/usr/include/boost/asio/ssl/impl/context.ipp:986: undefined reference to `ERR_clear_error'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:988: undefined reference to `SSL_CTX_use_PrivateKey_file'
/usr/bin/ld: /usr/include/boost/asio/ssl/impl/context.ipp:991: undefined reference to `ERR_get_error'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::engine(ssl_ctx_st*)':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:34: undefined reference to `SSL_new'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:39: undefined reference to `ERR_get_error'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:48: undefined reference to `SSL_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:49: undefined reference to `SSL_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:51: undefined reference to `SSL_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:55: undefined reference to `BIO_new_bio_pair'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:56: undefined reference to `SSL_set_bio'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::~engine()':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:61: undefined reference to `SSL_get_ex_data'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:63: undefined reference to `SSL_get_ex_data'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:64: undefined reference to `SSL_set_ex_data'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:67: undefined reference to `BIO_free'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:68: undefined reference to `SSL_free'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::get_output(boost::asio::mutable_buffer const&)':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:174: undefined reference to `BIO_read'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::put_input(boost::asio::const_buffer const&)':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:184: undefined reference to `BIO_write'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::map_error_code(boost::system::error_code&) const':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:199: undefined reference to `BIO_ctrl'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:213: undefined reference to `SSL_get_shutdown'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::perform(int (boost::asio::ssl::detail::engine::*)(void*, unsigned long), void*, unsigned long, boost::system::error_code&, unsigned long*)':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:233: undefined reference to `BIO_ctrl_pending'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:234: undefined reference to `ERR_clear_error'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:236: undefined reference to `SSL_get_error'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:237: undefined reference to `ERR_get_error'
/usr/bin/ld: /usr/include/boost/asio/ssl/detail/impl/engine.ipp:238: undefined reference to `BIO_ctrl_pending'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::do_accept(void*, unsigned long)':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:303: undefined reference to `SSL_accept'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::do_connect(void*, unsigned long)':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:308: undefined reference to `SSL_connect'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::do_read(void*, unsigned long)':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:321: undefined reference to `SSL_read'
/usr/bin/ld: /tmp/ccnxNuD8.o: in function `boost::asio::ssl::detail::engine::do_write(void*, unsigned long)':
/usr/include/boost/asio/ssl/detail/impl/engine.ipp:327: undefined reference to `SSL_write'
collect2: error: ld returned 1 exit status
```

Reply:
Probably you haven't linked with OpenSSL.

Reply:
you're right, the program compiles after linking SSL only to send an error upon request, will have to look into that.

Reply:
Probably it's due to missing key/cert files. Yeah, tests need some refactor too ;)

Reply:
I'm not sure about the tests but i did actually generate a self signed certificate (and set my rest client to not validate certificates), but the error seems to be related to the data transfer itself, most likely sockets again